docs/2023/puzzles/day14.md

```diff
@@ -2,10 +2,337 @@ import Solver from "../../../../../website/src/components/Solver.js"
 
 # Day 14: Parabolic Reflector Dish
 
+by [@anatoliykmetyuk](https://github.com/anatoliykmetyuk)
+
 ## Puzzle description
 
 https://adventofcode.com/2023/day/14
 
+## Part 1 Solution
+Part 1 of the puzzle features a square field where two types of objects can reside: round rocks 'O' and square rocks '#'. The objective is to "tilt" the field north, so that the round rocks roll as far as they can in that direction, while square rocks stay in place. Then, we need to compute the "loading" metric on the northern side of the field which is calculated as a function of how many round rocks there are on the field and how close they are to the northern edge.
+
+To solve Part 1, we do not even need to simulate the movement of the rocks - that is, we do not need to compute the state of the model of the platform after the tilting was performed. We can iterate from the top (north) of the field to the bottom, one line at a time, and use caching to remember where the rocks can fall. The algorithm is as follows:
+
+```scala
+def totalLoading(lines: List[String]): Int =
+  var loading = 0
+  val whereCanIFall = collection.mutable.Map.empty[Int, Int]
+  val totalRows = lines.size
+  for
+    (line, row) <- lines.zipWithIndex
+    (char, col) <- line.zipWithIndex
+  do char match
+    case 'O' =>
+      val fallRow = whereCanIFall.getOrElseUpdate(col, 0)
+      loading += totalRows - fallRow
+      whereCanIFall(col) = fallRow + 1
+    case '#' =>
+      whereCanIFall(col) = row + 1
+    case '.' =>
+  loading
+```
+
+## Part 2 Solution
+Part 2 is much more involved. Here, we are required to tilt the field in all four directions - North, West, South and East - in turn, in many cycles. The objective is to calculate the same loading metric after 1 billion cycles.
+
+The approach from Part 1 doesn't work here: we now need to modify our model state. The reason we didn't have to do it in Part 1 is that our model evolves in only a single step, and we only need one metric from the end state, so we don't have to calculate the entire final state. In Part 2, however, the model evolves in many steps, and each successive step depends on the previous step. So, we need to calculate the entire final state.
+
+### The Model
+Actually, we already know how to do the Northern tilt - similarly to Part 1. We do not want to re-implement the tilts for the other directions - instead, we want to use the Northern tilt as a building block for the other tilts. So, we need to be able to rotate our model in all four directions, and then tilt it North.
+
+We can represent the model as a 2D array of chars:
+
+```scala
+class Model(state: String):
+  private var dataN: Array[Array[Char]] = Array.empty
```

Comment:
maybe add a comment that the array is indexed as `dataN(x)(y)` for a given point (x, y)

Reply:
Done